Fix passing parameters to controller method with default parameter values

pako-pl opened the pull request:
For a route that looks like this:

    $app->get('/test/{param1}', 'TestController@test');

and a TestController method with a default parameter:

    public function test($param1 = 'default') {

visiting the URL "/test/param1Value" will cause Lumen to call the
method with the following parameters:

    TestController->test('default', 'param1Value');

The value in the URL will not replace $param1 default value, even
though the name in the route parameter is the same: {param1}.

The reason for that is that when this method is called:

    protected function addDependencyForCallParameter(ReflectionParameter $parameter, array &$parameters, &$dependencies)

the $parameters array contains only numeric keys. So this condition
in addDependencyForCallParameter() will never match:

    if (array_key_exists($parameter->name, $parameters))
    {
        $dependencies[] = $parameters[$parameter->name];
    
        unset($parameters[$parameter->name]);
    }

Removing array_values() call on the $parameters array in the callControllerAction() method seems to fix this issue.

taylorotwell commented:
Unit test trying to recreate the issue still fails even after merging this PR locally. Regardless, optional parameters are not even supported in Lumen so there is no need to ever have such an optional parameter.

pako-pl commented:
In our case removing `array_values()` helped, as the keys in the array were preserved. Parameter values can then be matched by name and the default value of `$param1` gets replaced. Why `array_values()` is even being called in the `callLumenController` method?

If you decide not to change the existing behaviour, then I think it's worth documenting. Currently parameters with optional values should not be used on controller methods, as the default value will never get replaced. Also there's no matching parameter names from the route to parameter names in the controller method - their order always needs to be the same.

For us, it's perfectly okay that default parameters values are not supported on the router level. But it would be useful to be able to have a default parameters value on the `performanceGraphs` method in cases like this:

    $app->get('/performance-graphs', 'TestController@performanceGraphsWithDefaultValue');
    $app->get('/performance-graphs/{graphsMode}', 'TestController@performanceGraphs');

Right now, it is necessary to add a separate `performanceGraphsWithDefaultValue()` method in the controller just to call `performanceGraphs()` with a default value.

pako-pl commented:
Just to sum up the existing behaviour:

- the parameters from the route are put in an ordered list
- Lumen scans the parameter list of the controller method and replaces parameter values with the values from the route (ignoring parameter names, only the parameter order is important); if there are class type hints it injects dependencies for that parameters instead
- if the controller method contains parameters with default values, the default value is always used; any remaining parameters from the route are appended after parameters with default values, which means they will be silently discarded

Taking a look at this method from `Illuminate\Container\Container`:

    protected function addDependencyForCallParameter(ReflectionParameter $parameter, array &$parameters, &$dependencies)
	{
		if (array_key_exists($parameter->name, $parameters))
		{
			$dependencies[] = $parameters[$parameter->name];

			unset($parameters[$parameter->name]);
		}
		elseif ($parameter->getClass())
		{
			$dependencies[] = $this->make($parameter->getClass()->name);
		}
		elseif ($parameter->isDefaultValueAvailable())
		{
			$dependencies[] = $parameter->getDefaultValue();
		}
	}

In Lumen's case it could be reduced to:

    protected function addDependencyForCallParameter(ReflectionParameter $parameter, array &$parameters, &$dependencies)
	{
		if ($parameter->getClass())
		{
			$dependencies[] = $this->make($parameter->getClass()->name);
		}
	}

It would be better because now it would only do dependency injection. Controller method parameters with default values would get their default values replaced with the values from the route, just like any other parameters. Any remaining parameters with default values will simply omitted from the method call, and their default values would be used.

taylorotwell commented:
Write a unit test to demonstrate your fix

On Monday, April 27, 2015, Paweł Kowalczyk <notifications@github.com> wrote:

> Just to sum up the existing behaviour:
>
>    - the parameters from the route are put in an ordered list
>    - Lumen scans the parameter list of the controller method and replaces
>    parameter values with the values from the route (ignoring parameter names,
>    only the parameter order is important); if there are class type hints it
>    injects dependencies for that parameters instead
>    - if the controller method contains parameters with default values,
>    the default value is always used; any remaining parameters from the route
>    are appended after parameters with default values, which means they will be
>    silently discarded
>
> Taking a look at this method from Illuminate\Container\Container:
>
> protected function addDependencyForCallParameter(ReflectionParameter $parameter, array &$parameters, &$dependencies)
> {
>     if (array_key_exists($parameter->name, $parameters))
>     {
>         $dependencies[] = $parameters[$parameter->name];
>
>         unset($parameters[$parameter->name]);
>     }
>     elseif ($parameter->getClass())
>     {
>         $dependencies[] = $this->make($parameter->getClass()->name);
>     }
>     elseif ($parameter->isDefaultValueAvailable())
>     {
>         $dependencies[] = $parameter->getDefaultValue();
>     }
> }
>
> In Lumen's case it could be reduced to:
>
> protected function addDependencyForCallParameter(ReflectionParameter $parameter, array &$parameters, &$dependencies)
> {
>     if ($parameter->getClass())
>     {
>         $dependencies[] = $this->make($parameter->getClass()->name);
>     }
> }
>
> It would be better because now it would only do dependency injection.
> Controller method parameters with default values would get their default
> values replaced with the values from the route, just like any other
> parameters. Any remaining parameters with default values will simply
> omitted from the method call, and their default values would be used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/lumen-framework/pull/54#issuecomment-96563495>
> .
>

pako-pl commented:
Could you please send me the unit test you used?

pako-pl commented:
Okay, now I've noticed that I missed array_values() calls in a few other places. I added another commit that removed them. Now it should work for callbacks too (earlier the fix applied only to controller methods). I also added a unit test:

https://github.com/pako-pl/lumen-framework/tree/fix-default-parameter-values

I'm not sure how to update this pull request as it's already closed. Should I open another one?

taylorotwell commented:
Open another one with unit tests.

On Mon, Apr 27, 2015 at 9:38 AM, Paweł Kowalczyk <notifications@github.com>
wrote:

> Okay, now I've noticed that I missed array_values() calls in a few other
> places. I added another commit that removed them. Now it should work for
> callbacks too (earlier the fix applied only to controller methods). I also
> added a unit test:
>
>
> https://github.com/pako-pl/lumen-framework/tree/fix-default-parameter-values
>
> I'm not sure how to update this pull request as it's already closed.
> Should I open another one?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/lumen-framework/pull/54#issuecomment-96679098>
> .
>
